apps/rpc/src/modules/event/attendance-service.ts

```diff
@@ -0,0 +1,795 @@
+        )
+      }
+
+      // Marking the attendee as registered is only half of the job, as we also schedule a task to reserve their place
+      // in the pool
+      let reservationTime = addHours(getCurrentUTC(), applicablePool.mergeDelayHours ?? 0)
+      if (punishment !== null) {
+        reservationTime = addHours(reservationTime, punishment.delay)
+      }
+      const attendee = await attendanceRepository.createAttendee(handle, attendanceId, applicablePool.id, userId, {
+        attendedAt: null,
+        earliestReservationAt: reservationTime,
+        reserved: options.immediateReservation,
+        selections: [],
+        userGrade: grade,
+      })
+
+      if (attendance.attendancePrice) {
+        const paymentDeadline = options.immediatePayment ? addMinutes(new TZDate(), 15) : addHours(new TZDate(), 24)
+        await this.createAttendeePayment(handle, attendee.id, paymentDeadline)
+      }
+
+      // When a user is immediately reserved, there is no reason to schedule a task for them.
+      if (!options.immediateReservation) {
+        await taskSchedulingService.scheduleAt(
+          handle,
+          tasks.RESERVE_ATTENDEE,
+          {
+            attendeeId: attendee.id,
+            attendanceId,
+          },
+          reservationTime
+        )
+      }
+      return attendee
+    },
+    async updateAttendeeById(handle, attendeeId, data) {
+      const attendance = await this.getAttendanceByAttendeeId(handle, attendeeId)
+      const attendee = attendance.attendees.find((attendee) => attendee.id === attendeeId)
+      invariant(attendee !== undefined)
+      // If the user is not suspended, we can update the attendee.
+      const input = {
+        ...attendee,
+        ...data,
+      } satisfies AttendeeWrite
+      validateAttendeeWrite(input)
+      return await attendanceRepository.updateAttendeeById(handle, attendeeId, input)
+    },
+    async executeReserveAttendeeTask(handle, { attendanceId, attendeeId }) {
+      const attendance = await this.getAttendanceById(handle, attendanceId)
+      const attendee = attendance.attendees.find((a) => a.id === attendeeId)
+      // NOTE: If the attendee does not exist, we have a non-critical bug in the app. The circumstances where this is
+      // possible is when the attendee was removed from the attendance after the task was scheduled AND the task was not
+      // cancelled.
+      if (attendee === undefined) {
+        throw new AttendanceNotFound(`Attendee(ID=${attendeeId}) not found in Attendance(ID=${attendanceId})`)
+      }
+      if (attendee.reserved) {
+        return
+      }
+
+      const pool = attendance.pools.find((pool) => pool.id === attendee.attendancePoolId)
+      invariant(pool !== undefined)
+      const adjacentAttendees = attendance.attendees.filter((a) => a.attendancePoolId === pool.id)
+      const isPoolAtMaxCapacity = adjacentAttendees.length === pool.capacity
```

Comment:
there is a real case where you have more attendees than there is capacity if you admin register people
```suggestion
      const adjacentAttendees = attendance.attendees.filter((a) => a.attendancePoolId === pool.id && a.reserved)
      const isPoolAtMaxCapacity = adjacentAttendees.length >= pool.capacity
```